timely_beliefs/beliefs/queries.py

```diff
@@ -0,0 +1,60 @@
+from datetime import timedelta
+from typing import Optional
+
+from sqlalchemy import and_, func, or_
+from sqlalchemy.orm import Query, Session, aliased
+
+
+def query_unchanged_beliefs(
+    session: Session,
+    cls: "TimedBeliefDBMixin",  # noqa F821
+    query: Optional[Query] = None,
+    include_positive_horizons: bool = True,  # include belief horizons > 0 (i.e. forecasts)
+    include_non_positive_horizons: bool = True,  # include belief horizons <= 0 (i.e. measurements, nowcasts and backcasts)
+) -> Query:
+    """Match unchanged beliefs.
+
+    Unchanged beliefs are beliefs that have not changed with respect to the preceding belief,
+    other than their belief time.
+    """
+    if query is None:
+        query = cls.query
+
+    # Set up aliases
+    tb1 = cls  # the DBTimedBelief class mapped to the timed_beliefs table
+    tb2 = aliased(cls)  # alias for holding each preceding belief
+    tb3 = aliased(cls)  # alias from which to select the preceding belief
+
+    # Set up / copy criteria for the tb3 alias
+    # todo: if query has filter criteria on tb1, those should be applied to tb2 and tb3, too. Hint, combine visitors.iterate(query.whereclause) with replace_selectable
+    tb3_criteria = []
+    if include_positive_horizons:
+        tb3_criteria.append(tb3.belief_horizon > timedelta(0))
+    if include_non_positive_horizons:
+        tb3_criteria.append(tb3.belief_horizon <= timedelta(0))
+
+    q = query.join(
+        tb2,
+        and_(
+            tb2.event_start == tb1.event_start,
+            tb2.sensor_id == tb1.sensor_id,
+            tb2.source_id == tb1.source_id,
+            # next higher belief horizon for a given event for a given source, i.e. the preceding belief
+            tb2.belief_horizon
+            == session.query(func.min(tb3.belief_horizon))
+            .where(
+                tb3.belief_horizon > tb1.belief_horizon,
+                tb3.event_start == tb1.event_start,
+                tb3.sensor_id == tb1.sensor_id,
+                tb3.source_id == tb1.source_id,
+                *tb3_criteria,
+            )
+            .scalar_subquery(),
+        ),
+    ).where(
+        or_(tb2.source_id == tb1.source_id, tb2.source_id is None),
```

Comment:
Why this again? It's also in line 41

Reply:
Thanks. This made me realize I could simplify a bit further.

Historical note: I had added this line to support the opposite query (getting changed beliefs), in which the case the very oldest belief in tb1 has all None values in tb2.